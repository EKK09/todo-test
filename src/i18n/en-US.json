{
    "global": {
        "appName": "TODO LIST",
        "logout": "Logout"
    },
    "loginPage": {
        "account": "Account",
        "password": "Password",
        "login": "Login"
    },
    "todo": {
        "completed": "Completed"
    }
}