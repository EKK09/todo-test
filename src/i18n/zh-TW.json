{
    "global": {
        "appName": "代辦事項"
    },
    "loginPage": {
        "account": "帳號",
        "password": "密碼",
        "login": "登入"
    },
    "todo": {
        "completed": "完成"
    }
}