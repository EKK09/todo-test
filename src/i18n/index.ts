import i18n from 'i18next';
import { initReactI18next } from 'react-i18next';
import enUSResource from './en-US.json';
import zhTWResource from './zh-TW.json';

export enum Lang {
    EN_US = 'en-US',
    ZH_TW = 'zh-TW',
}

export const LANG_OPTION = {
  [Lang.EN_US]: 'English',
  [Lang.ZH_TW]: '中文',
};

const resources = {
  [Lang.EN_US]: {
    translation: enUSResource,
  },
  [Lang.ZH_TW]: {
    translation: zhTWResource,
  },
};

const getDefaultLang = (locale: string): Lang => {
  switch (locale) {
    case Lang.EN_US:
      return Lang.EN_US;

    case Lang.ZH_TW:
      return Lang.ZH_TW;

    default:
      return Lang.EN_US;
  }
};

const locale = window.navigator.language;

i18n
  .use(initReactI18next)
  .init({
    resources,
    lng: getDefaultLang(locale),
  });

export default i18n;
