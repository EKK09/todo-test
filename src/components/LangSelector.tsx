import React, { ChangeEvent } from 'react';
import { LANG_OPTION } from 'src/i18n';
import styled from 'styled-components';
import { useTranslation } from 'react-i18next';

const Wrapper = styled.div`
  margin: 0 16px;
  display: flex;
  justify-content: space-between;
  align-items: center;
`;
const Select = styled.select`
    font-size: 1em;
`;

const LangSelector = (): React.ReactElement => {
  const { i18n } = useTranslation();
  const currentlang = i18n.language;
  const handleLangSelect = (event: ChangeEvent<HTMLSelectElement>) => {
    i18n.changeLanguage(event.target.value);
  };
  const langOptions = Object.entries(LANG_OPTION).map(([value, label]) => (
    <option key={value} value={value}>{label}</option>
  ));
  return (
    <Wrapper>
      <Select value={currentlang} onChange={handleLangSelect} data-testid="lang-selector">
        {langOptions}
      </Select>
    </Wrapper>
  );
};

export default LangSelector;
