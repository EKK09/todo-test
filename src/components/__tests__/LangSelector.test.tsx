import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { I18nextProvider } from 'react-i18next';
import i18n, { Lang } from 'src/i18n';
import LangSelector from '../LangSelector';

describe('LangSelector component', () => {
  it('Render LangSelector', async () => {
    i18n.changeLanguage(Lang.EN_US);
    const wrapper = render(
      <I18nextProvider i18n={i18n}>
        <LangSelector />
      </I18nextProvider>,
    );

    expect(wrapper.container).toMatchSnapshot();
  });

  it('selector onchange should call changeLanguage ', async () => {
    const mockChangeLanguage = jest.fn();
    i18n.changeLanguage = mockChangeLanguage;
    const { getByTestId } = render(
      <I18nextProvider i18n={i18n}>
        <LangSelector />
      </I18nextProvider>,
    );
    const selectNode = getByTestId('lang-selector');
    fireEvent.change(selectNode, { target: { value: Lang.ZH_TW } });

    expect(mockChangeLanguage).toBeCalledWith(Lang.ZH_TW);
  });
});
