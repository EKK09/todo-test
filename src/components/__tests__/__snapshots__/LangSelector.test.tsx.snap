// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LangSelector component Render LangSelector 1`] = `
<div>
  <div
    class="sc-bdfBQB kfrCnO"
  >
    <select
      class="sc-gsTEea gNqOUL"
      data-testid="lang-selector"
    >
      <option
        value="en-US"
      >
        English
      </option>
      <option
        value="zh-TW"
      >
        中文
      </option>
    </select>
  </div>
</div>
`;
